### PR TITLE
query network version when metamask is back on line

### PR DESCRIPTION
**Pull request by frankiebee**
partial fix for issue #2636 however their maybe some other things in that issue that i am not considering.

the issue this fixes is that if a provider was created when offline the network spinner goes for ever and a request never goes through for net_version, this should fix that. 

their are subsequent issues that have been discovered in the pursuit of fixing this one:
https://github.com/MetaMask/eth-block-tracker/issues/27
https://github.com/MetaMask/metamask-extension/issues/3003

### CODE REVIEW

File: app/scripts/controllers/network.js [hunk 1]
```diff
@@ -58,15 +58,20 @@ module.exports = class NetworkController extends EventEmitter {
...
     this.ethQuery.sendAsync({ method: 'net_version' }, (err, network) => {
-      if (err) return this.setNetworkState('loading')
+      if (err) {
+        this.setNetworkState('loading')
+        console.error(err)
+        return
+      }
       log.info('web3.getNetwork returned ' + network)
       this.setNetworkState(network)
+      cb(err, network)
```
Comment (kumavis) [inline app/scripts/controllers/network.js]: if `err` is truthy, this part will not be reached

File: app/scripts/controllers/network.js [hunk 2]
```diff
@@ -58,15 +58,20 @@ module.exports = class NetworkController extends EventEmitter {
...
     // Prevent firing when provider is not defined.
     if (!this.ethQuery || !this.ethQuery.sendAsync) {
       return log.warn('NetworkController - lookupNetwork aborted due to missing ethQuery')
     }
     this.ethQuery.sendAsync({ method: 'net_version' }, (err, network) => {
-      if (err) return this.setNetworkState('loading')
+      if (err) {
+        this.setNetworkState('loading')
+        console.error(err)
+        return
```
Comment (kumavis) [inline app/scripts/controllers/network.js]: `cb` not called

File: app/scripts/controllers/network.js [hunk 3]
```diff
@@ -58,15 +58,20 @@ module.exports = class NetworkController extends EventEmitter {
     return this.getNetworkState() === 'loading'
   }
 
-  lookupNetwork () {
+  lookupNetwork (cb = () => {}) {
```
Comment (kumavis) [inline app/scripts/controllers/network.js]: define `noop` up by the require statements

File: app/scripts/platforms/extension.js [hunk 4]
```diff
@@ -26,6 +26,14 @@ class ExtensionPlatform {
       cb(e)
     }
   }
+
+  isOnline () {
+    return navigator.onLine
+  }
+
+  onLine (cb) {
```
Comment (kumavis) [inline app/scripts/platforms/extension.js]: not clearly named

File: app/scripts/metamask-controller.js [hunk 5]
```diff
@@ -61,6 +61,21 @@ module.exports = class MetamaskController extends EventEmitter {
     // network store
     this.networkController = new NetworkController(initState.NetworkController)
 
+    // setup onLine lister for restarting the provider
+    if (this.platform.onLine) {
+      this.platform.onLine(() => {
+        // guard against premature firring of the event listener
+        if (this.provider) {
+          // start the provider up
+          this.provider.stop()
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: this is referring to the blockTracker actually, correct?

File: app/scripts/metamask-controller.js [hunk 6]
```diff
@@ -61,6 +61,21 @@ module.exports = class MetamaskController extends EventEmitter {
...
+          this.provider.start()
+        }
+        // look up the network
+        this.networkController.lookupNetwork(() => {
+          this.emit('update', this.getState())
+        })
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: this is not the correct way to emit updates for MetamaskController

**Review by kumavis — DISMISSED** [review]
nice find, implementation needs some improvements

**Comment by danfinlay** [comment]
Related to #2548

File: app/scripts/controllers/network.js [hunk 7]
```diff
@@ -58,15 +59,21 @@ module.exports = class NetworkController extends EventEmitter {
...
+  lookupNetwork (cb = noop){
     // Prevent firing when provider is not defined.
     if (!this.ethQuery || !this.ethQuery.sendAsync) {
       return log.warn('NetworkController - lookupNetwork aborted due to missing ethQuery')
     }
     this.ethQuery.sendAsync({ method: 'net_version' }, (err, network) => {
-      if (err) return this.setNetworkState('loading')
+      if (err) {
+        this.setNetworkState('loading')
+        console.error(err)
```
Comment (danfinlay) [inline app/scripts/controllers/network.js]: Should be `log.error` to use our `loglevel` utility.

File: app/scripts/metamask-controller.js [hunk 8]
```diff
@@ -61,6 +61,21 @@ module.exports = class MetamaskController extends EventEmitter {
     // network store
     this.networkController = new NetworkController(initState.NetworkController)
 
+    // setup onLine lister for restarting the provider
+    if (this.platform.addOnLineListener) {
```
Comment (danfinlay) [inline app/scripts/metamask-controller.js]: "online" is one word, I'd rather we didn't camel case it into two. Non blocking.

File: app/scripts/platforms/extension.js [hunk 9]
```diff
@@ -26,6 +26,14 @@ class ExtensionPlatform {
       cb(e)
     }
   }
+
+  isOnline () {
+    return navigator.onLine
+  }
+
+  addOnLineListener (cb) {
+    window.addEventListener('online', cb)
```
Comment (danfinlay) [inline app/scripts/platforms/extension.js]: Yeah camel cased correctly here. Incorrect capitalization kinda driving me nuts, going to change it myself.

**Review by danfinlay — CHANGES_REQUESTED** [review]
Also should have a changelog.

Frankie isn't doing well, so I'm going to make my requested changes.

**Comment by danfinlay** [comment]
I can readily reproduce an instance of #2636 on this branch with these steps:

1. Select main network
2. Ensure Ganache/local network is not running
3. Switch to the local network which is not running
4. Turn on the local network/ganache

Results: The network spinner stays on.

The block tracker's `start` method returns a promise that does not seem to get any retries, so only if the disconnect is 

## Regression:

If a test network is successfully connected, and then disconnected, the network indicator is never shown. This seems to me like it breaks the basic functionality of the network connection spinner, and is arguably worse than the current behavior.

I wouldn't want to merge this until the spinner accurately reflects network status.

File: app/scripts/platforms/extension.js [hunk 10]
```diff
@@ -26,6 +26,14 @@ class ExtensionPlatform {
       cb(e)
     }
   }
+
+  isOnline () {
+    return navigator.onLine
```
Comment (kumavis) [inline app/scripts/platforms/extension.js]: sadly, this is the webapi's camelization

**Comment by kumavis** [comment]
@frankiebee said we can retire this one

**Pull request closed without merging**